### PR TITLE
Transition __attribute__((constructor)) to a more platform independent approach

### PR DESCRIPTION
**Pull request by artemdinaburg**
These commits remove the use of `__attribute__((constructor))` in favor of a more platform independent construct using static initializers. These changes are needed since there is no constructor attribute on Windows. This is needed to get a working shell and daemon #2007 

This code passes tests, but it could use testing with more external modules.

### CODE REVIEW

**Comment by osqueryer** [comment]
Can one of the osquery admins reply with "ok to test" to kick off a build...

**Comment by PoppySeedPlehzr** [comment]
ok to test

**Comment by osqueryer** [comment]
Build origin/pr/2233/merge (Job: 2852) is good! :+1:

File: include/osquery/sdk.h [hunk 1]
```diff
@@ -54,15 +54,14 @@ REGISTER_INTERNAL(ExternalSQLPlugin, "sql", "sql");
...
-/// Helper macro to write the `initModule` symbol without rewrites.
-#define DECLARE_MODULE(name)        \
-  extern "C" void initModule(void); \
-  __attribute__((constructor)) static void declareModule()
+#define REGISTER_EXTERNAL(type, registry, name)                               \
+  struct type##ExtensionRegistryItem : public InitializerInterface {          \
+    type##ExtensionRegistryItem(void) { registerPlugin(this); }               \
+    const char *id() const override { return name_.c_str(); }                 \
+    void run() const override { Registry::add<type>(registry, name); }        \
+    std::string name_{std::string(registry) + "." + name};                    \
```
Comment (theopolis) [inline include/osquery/sdk.h]: Can we move this into a preprocessor-concat `const char*` within the `id()` return?

File: osquery/core/init.cpp [hunk 2]
```diff
@@ -195,6 +199,44 @@ volatile std::sig_atomic_t kExitCode{0};
...
+  }
+}
+
+void beginRegistryAndPluginInit() {
+  for (auto it = registry_initializer().begin();
+       it != registry_initializer().end(); ++it) {
+    it->second->run();
+  }
+
+  for (auto it = plugin_initializer().begin(); it != plugin_initializer().end();
```
Comment (theopolis) [inline osquery/core/init.cpp]: Does:
```cpp
for (auto it : plugin_initializer()) {

}
```
not work?

File: osquery/core/init.cpp [hunk 3]
```diff
@@ -195,6 +199,44 @@ volatile std::sig_atomic_t kExitCode{0};
...
+
+void registerPlugin(InitializerInterface *const item) {
+  if (item != nullptr) {
+    plugin_initializer().insert(
+        std::pair<std::string, InitializerInterface *>(item->id(), item));
+  }
+}
+
+void beginRegistryAndPluginInit() {
+  for (auto it = registry_initializer().begin();
```
Comment (theopolis) [inline osquery/core/init.cpp]: Same as below for iteration.

File: osquery/core/init.cpp [hunk 4]
```diff
@@ -195,6 +199,44 @@ volatile std::sig_atomic_t kExitCode{0};
...
+
+InitializerMap& plugin_initializer() {
+  static InitializerMap plugin_;
+  return plugin_;
+}
+
+void registerRegistry(InitializerInterface *const item) {
+  if (item != nullptr) {
+    registry_initializer().insert(
+        std::pair<std::string, InitializerInterface *>(item->id(), item));
```
Comment (theopolis) [inline osquery/core/init.cpp]: Can you write `std::make_pair` instead?

Comment (zwass) [inline osquery/core/init.cpp]: I think it can even be just `{item->id(), item}`.

**Comment by osqueryer** [comment]
Build origin/pr/2233/merge (Job: 2856) is good! :+1:

**Pull request merged**